ibis/expr/datatypes.py

```diff
@@ -189,144 +199,178 @@ class Time(Primitive):
     __slots__ = ()
 
 
-class Timestamp(Primitive):
+class Timestamp(DataType):
     scalar = ir.TimestampScalar
     column = ir.TimestampColumn
 
     __slots__ = 'timezone',
 
-    def __init__(self, timezone=None, nullable=True):
-        super(Timestamp, self).__init__(nullable=nullable)
+    def __init__(
+        self,
+        timezone: Optional[str] = None,
+        nullable: bool = True
+    ) -> None:
+        super().__init__(nullable=nullable)
         self.timezone = timezone
 
-    def _equal_part(self, other, cache=None):
-        return self.timezone == other.timezone
-
-    def __call__(self, timezone=None, nullable=True):
-        return type(self)(timezone=timezone, nullable=nullable)
-
-    def __str__(self):
+    def __str__(self) -> str:
         timezone = self.timezone
         typename = self.name.lower()
         if timezone is None:
             return typename
         return '{}({!r})'.format(typename, timezone)
 
-    def __repr__(self):
-        return DataType.__repr__(self)
+
+class BoundedInteger(Integer):
+    __slots__ = 'nbytes', 'bounds'
+
+    def __init__(
+        self,
+        nbytes: int,
+        bounds: IntegerBounds,
+        nullable: bool = True
+    ) -> None:
+        super().__init__(nullable=nullable)
+        self.nbytes = nbytes
+        self.bounds = bounds
 
 
 class SignedInteger(Integer):
+    __slots__ = ()
 
-    @property
-    def largest(self):
-        return int64
 
+class BoundedSignedInteger(BoundedInteger, SignedInteger):
+    __slots__ = ()
 
-class UnsignedInteger(Integer):
+    def __init__(self, nbytes: int, nullable: bool = True) -> None:
+        super().__init__(
+            nbytes=nbytes,
+            bounds=IntegerBounds(
+                lower=-1 << (nbytes * 8 - 1),
+                upper=~(-1 << (nbytes * 8 - 1))
+            ),
```

Comment:
What was wrong with the previous implementation? It seems like overcomplicated.

Reply:
I will revert this back to the original and make a follow-up PR.

Reply:
Thanks :)